grpcwebclient/impl.ts

```diff
@@ -0,0 +1,296 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { ConformanceServiceClient } from "./gen/proto/connectrpc/conformance/v1/ServiceServiceClientPb.js";
+import {
+  ClientCompatRequest,
+  ClientResponseResult,
+} from "./gen/proto/es/connectrpc/conformance/v1/client_compat_pb.js";
+import {
+  ConformancePayload,
+  Error as ProtoError,
+  Header,
+  UnaryRequest,
+  UnimplementedRequest,
+  ServerStreamRequest,
+} from "./gen/proto/es/connectrpc/conformance/v1/service_pb.js";
+import {
+  UnaryRequest as UnaryRequestGoog,
+  UnaryResponse as UnaryResponseGoog,
+  UnimplementedRequest as UnimplementedRequestGoog,
+  ServerStreamRequest as ServerStreamRequestGoog,
+  ServerStreamResponse as ServerStreamResponseGoog,
+} from "./gen/proto/connectrpc/conformance/v1/service_pb.js";
+import { Status } from "@buf/googleapis_googleapis.bufbuild_es/google/rpc/status_pb.js";
+import { Metadata, RpcError, Status as GrpcWebStatus } from "grpc-web";
+
+function convertGrpcToProtoError(rpcErr: RpcError): ProtoError {
+  const err = new ProtoError({
+    code: rpcErr.code,
+    message: rpcErr.message,
+  });
+  for (const [name, value] of Object.entries(rpcErr.metadata)) {
+    if (name === "grpc-status-details-bin") {
+      const status = Status.fromBinary(stringToUint8Array(atob(value)));
+      err.details = status.details;
+      break;
+    }
+  }
+  return err;
+}
+
+function stringToUint8Array(str: string): Uint8Array {
+  const buf = new ArrayBuffer(str.length);
+  const bufView = new Uint8Array(buf);
+  for (let i = 0; i < str.length; i++) {
+    bufView[i] = str.charCodeAt(i);
+  }
+  return bufView;
+}
+
+function buildMetadata(req: ClientCompatRequest): Metadata {
+  const metadata: Metadata = {};
+  req.requestHeaders.forEach((hdr: Header) => {
+    const s = hdr.value.join(",");
+    metadata[hdr.name] = s;
+  });
+
+  if (req.timeoutMs !== undefined && req.timeoutMs > 0) {
+    let deadline = new Date();
+    deadline.setMilliseconds(deadline.getMilliseconds() + req.timeoutMs);
+    metadata.deadline = deadline.getTime().toString();
+  }
+  return metadata;
+}
+
+function convertMetadataToHeader(md: Metadata): Header[] {
+  const hdrs: Header[] = [];
+  for (const [name, value] of Object.entries(md)) {
+    hdrs.push(
+      new Header({
+        name,
+        value: [value],
+      }),
+    );
+  }
+
+  return hdrs;
+}
+
+async function unary(
+  client: ConformanceServiceClient,
+  req: ClientCompatRequest,
+): Promise<ClientResponseResult> {
+  const msg = req.requestMessages[0];
+  const uReq = new UnaryRequest();
+  if (!msg.unpackTo(uReq)) {
+    throw new Error("Could not unpack request message to unary request");
+  }
+
+  // Convert from Protobuf-ES into the gRPC-web compatible library
+  const ur = UnaryRequestGoog.deserializeBinary(uReq.toBinary());
```

Comment:
re: above comment about using both plugins -- this is where I convert the Protobuf-ES requests to the `protocolbuffers/js` requests. I also hate the `Goog` suffix, so if there's a better way to name them, lmk.